app/src/main/java/com/example/covidsafe/ui/PermissionLogic.java

```diff
@@ -22,22 +24,66 @@
 import com.google.android.material.dialog.MaterialAlertDialogBuilder;
 
 public class PermissionLogic {
+    @RequiresApi(api = Build.VERSION_CODES.Q)
     public static void permissionLogic(int requestCode, String[] permissions, int[] grantResults, final Activity av) {
+        int androidSDKVersion = Build.VERSION.SDK_INT;
+
         Log.e("logme","on request permission "+requestCode);
         for (int i = 0; i < grantResults.length; i++) {
             Log.e("logme","grant results "+permissions[i]+","+grantResults[i]);
         }
 
-        if (ActivityCompat.checkSelfPermission(av, Manifest.permission.ACCESS_BACKGROUND_LOCATION) == PackageManager.PERMISSION_DENIED) {
-            boolean shouldAsk = ActivityCompat.shouldShowRequestPermissionRationale(av, Manifest.permission.ACCESS_BACKGROUND_LOCATION);
+        if (androidSDKVersion >= Build.VERSION_CODES.Q && ActivityCompat.checkSelfPermission(av, Manifest.permission.ACCESS_BACKGROUND_LOCATION) == PackageManager.PERMISSION_DENIED) {
+            boolean shouldAsk;
+            shouldAsk = ActivityCompat.shouldShowRequestPermissionRationale(av, Manifest.permission.ACCESS_BACKGROUND_LOCATION);
+
             Log.e("logme","does not have permissions for tracking "+shouldAsk);
             if (requestCode == 1 && shouldAsk) {
                 AlertDialog dialog = new MaterialAlertDialogBuilder(av)
                         .setTitle("Permission denied")
                         .setMessage(av.getString(R.string.perm_ble_rationale))
                         .setNegativeButton(R.string.retry, new DialogInterface.OnClickListener() {
                             public void onClick(DialogInterface dialog, int which) {
-                                ActivityCompat.requestPermissions(av, new String[]{Manifest.permission.ACCESS_BACKGROUND_LOCATION}, 2);
+                                if (Build.VERSION.SDK_INT == Build.VERSION_CODES.Q) {
+                                    ActivityCompat.requestPermissions(av, new String[]{Manifest.permission.ACCESS_BACKGROUND_LOCATION}, 2);
+                                }
+                                else {
+                                    ActivityCompat.requestPermissions(av, new String[]{Manifest.permission.ACCESS_BACKGROUND_LOCATION}, 2);
+                                }
```

Comment:
Why is the same command in both the `if` and `else` path?  

I'm not familiar with Android permissions but I assume there was some change between Android Q and the prior versions, in which case it should reflect here. Otherwise, we don't need the branch.